Add support for emitting events from stack values

adr1anh opened the pull request:
## Describe your changes

Fixes #2068

This PR changes the semantic of `emit` by reading the event ID from the stack rather than from the immediate value in the instruction. As a consequence, `emit.ID` now gets lowered to a push instruction followed by an emit, taking 2 cycles instead of one. 

Due to the current implementation of `SlowProcessorState`, we also change when the clock is advanced when handling an event. Previously, the clock would increase *after* executing the event, whereas we now increment if *before*. This ensures that handlers have access to the full stack without the event ID at the top. 

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Commits are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.
- Updated `CHANGELOG.md'

plafer commented on processor/src/fast/mod.rs:
```diff
@@ -891,10 +891,21 @@ impl FastProcessor {
+                Operation::Emit => {
+                    let event_id = self.stack_get(0).as_int() as u32;
+
+                    // The event ID is removed from the stack before calling handlers via op_emit.
+                    // For consistency with SlowProcessor, we also increment the clock cycle before
+                    // executing the emit operation.
+                    self.decrement_stack_size();
+                    self.clk += 1_u32;
```
I understand why you went with this solution, but I don't think it's a viable one, since the user gets the incorrect clock cycle in the event handler (i.e. off by one, pointing to the next operation). Looking a bit more into it, the root of the issue is that `ProcessState` is designed to return the stack state *at the start* of the clock cycle before the operation is executed (tailored to use cases like debugger, where you want the debugger to stop and observe the state before the selected operation is executed), whereas in this case we want the state of the stack *after* the operation is executed. Hence your solution to return the state at the beginning of the following clock cycle.

However, thinking about this more, I'm starting to think it would be better to revisit the idea of making `EMIT` a degree-2 "no stack shift" operation.

As I understand it, the benefits are:
- free up a high degree operation slot
- the handlers are still "as powerful" as before, since they get access to the entire stack, not just the top 16 elements
  - I initially thought during the call that they only had access to the top 16 elements of the stack, so we'd be restraining them from using one of those slots - but this is incorrect

The downsides are:
- we break all existing handlers
- pretty much every `emit` will be followed by a `drop` (hence taking an extra cycle)

During the call, it seems like we favored not breaking existing handlers in favor of freeing a high degree operation slot - but I no longer believe this is the right call. Especially since fixing the handlers should be rather trivial, and we're still "allowed" to make such breaking changes.

Finally, circling back to the issue here, if `emit` didn't touch the stack, we'd get the added benefit that it would be simpler to implement correctly (i.e. no `clk` hack as is done here).

adr1anh replied:
I agreed with @plafer's comment about modifying the clock. It was the easiest "fix" compared to what I was [doing before](https://github.com/0xMiden/miden-vm/commit/5985fe05599fa06d930ee669edbe5a94384f5495). The issue was really with the slow processor, since we could easily decrement the stack in the fast processor before calling the event handler. From what I can tell, the checkpoint PR would be compatible with this change, though I may be missing something.

I also agree with @bobbinth's comment that handlers should have access to the full stack, since they're only used to modify the advice state. From a quick glance, we're only using the `ProcessState` for handling events (trace, debug, `op_assert`, and system/user event handlers). All of these should be allowed to access the full stack. 

For now and in the interest of time, moving `EMIT` to another bucket would take more time than leaving the hack in place for the slow processor. I don't have the full picture around our current constraints and how to modify them safely. In particular, I'd revert the changes from the linked commit but clean it up to avoid allocating the full stack vector at every stack access.

plafer replied:
> I think my main concern here is not the handlers that we control, but that any custom event would now need to account for an "extra" value being on top of the stack that should be ignored in basically all cases.

I wouldn't say it's useless - we already provide that information [here](https://github.com/0xMiden/miden-vm/blob/e961d413e40f8380ec30e00451923ecb3d0128c3/processor/src/host/mod.rs#L110). In the new version, it would be on top of the stack instead of being passed as a parameter to `on_event()`. Therefore that value has a purpose IMO (i.e. reusing the same handlers for multiple event IDs), even if we don't currently make use of it. I don't see our current handlers simply ignoring it as a downside - especially given the benefits that making emit a no stack shift operation provides.

> But I wonder if there is a way to have our cake and eat it too. Specifically, maybe the ProcessState struct can account for this and basically ignore the top element of the stack.

I think this would radically increase the complexity of event handling, and for a reason I don't understand. I prefer the value to be *explicit* on top of the stack, and explicitly ignored by handlers that don't need it, than us trying to hide it somehow.

> I also agree with @bobbinth's comment that handlers should have access to the full stack, since they're only used to modify the advice state.

They currently *do* have access to the whole stack (i.e. stack top + overflow) in both cases - what am I missing?

> So, to summarize: assuming this "fiddling" with ProcessState can be done relatively quickly, I think we can change the emit instruction which gets converted to push.EVENT_ID emit drop sequence of operations.

I don't think we should fiddle with `ProcessState` as described above, but I do agree that we should make `emit` a no stack shift. I'm okay with leaving it in the high degree bucket for now in the interest of time - it'll be simple to move it to the "no stack shift" bucket later *if it already has the no stack shift semantics*. @adr1anh your latest commit still has `emit` drop the element - given all the above, I believe it should
1. keep its opcode for now
2. not shift the stack

bobbinth replied:
> I wouldn't say it's useless - we already provide that information [here](https://github.com/0xMiden/miden-vm/blob/e961d413e40f8380ec30e00451923ecb3d0128c3/processor/src/host/mod.rs#L110). In the new version, it would be on top of the stack instead of being passed as a parameter to `on_event()`. Therefore that value has a purpose IMO (i.e. reusing the same handlers for multiple event IDs), even if we don't currently make use of it. I don't see our current handlers simply ignoring it as a downside - especially given the benefits that making emit a no stack shift operation provides.

Ah - good point! Then it is quite a bit simpler than I thought. Basically, `on_event()` handler doesn't need to take the `event_id` as a separate parameter and can read it as the first element from the stack. I think this could work pretty nicely.

adr1anh commented:
I've gone ahead and reverted the change that incremented the clock _before_ executing the event. It improves on the previous hack where we would get the full stack state whenever we read an item. I also went ahead and allow reading the stack beyond the minimum depth.

This should be good for a final review.

bobbinth commented:
@adr1anh - how difficult would be to be update this PR to reduce make `emit` a non-stack-shifting operation (while still keeping it in the high-degree bucket)? Basically, we'd need to do the following:

First, transform `emit` to `emit drop` and `emit.imm` into `push(immi) emit drop`.

Then, update the `Host::on_event()` handler to be to look as follows:
```rust
fn on_event(&mut self, process: &ProcessState) -> Result<Vec<AdviceMutation>, EventError>;
```

Then, update the existing event handlers to account for `event_id` being on top of the stack.

Thinking more about this, I think this is probably the long-term solution that should avoid the hacks - so, maybe good to do it sooner rather than later (and moving `emit` to a low-degree bucket can be done at some later point in time).

bobbinth commented on docs/src/design/decoder/constraints.md:
```diff
@@ -10,21 +10,21 @@ We assume that the VM exposes a flag per operation which is set to $1$ when the
+| `END`     |   $f_{end}$   |   4    | When exiting a loop block, top stack element is dropped; otherwise, the stack remains unchanged. |
+| `HALT`    |  $f_{halt}$   |   4    | Stack remains unchanged.                                                                         |
+| `PUSH`    |  $f_{push}$   |   5    | An immediate value is pushed onto the stack.                                                     |
+| `EMIT`    |  $f_{emit}$   |   5    | Top stack element is dropped.                                                                    |
```
I think the effect on the stack here is still "Stack remains unchanged."

bobbinth commented on docs/src/design/decoder/constraints.md:
```diff
@@ -628,10 +628,10 @@ Where $i \in [1, 8)$. Thus, $v_1$ defines row value for group in $h_1$, $v_2$ de
 
 $$
-u = \alpha_0 + \alpha_1 \cdot a + \alpha_2 \cdot gc + \alpha_3 \cdot ((h_0' \cdot 2^7 + op') \cdot (1 - f_{imm}) + s_0' \cdot f_{push} + h_2 \cdot f_{emit}) \text{ | degree} = 6
+u = \alpha_0 + \alpha_1 \cdot a + \alpha_2 \cdot gc + \alpha_3 \cdot ((h_0' \cdot 2^7 + op') \cdot (1 - f_{imm}) + s_0' \cdot f_{push}) \text{ | degree} = 6
```
Related to one of the previous comments: could we check if the degree here stays the same or goes up by $1$?

bobbinth commented on processor/src/stack/mod.rs:
```diff
@@ -161,13 +161,38 @@ impl Stack {
     /// Creating a word does not change the state of the stack.
-    pub fn get_word(&self, word_idx: usize) -> Word {
-        let offset = word_idx * WORD_SIZE;
+    pub fn get_word(&self, idx: usize) -> Word {
+        [self.get(idx + 3), self.get(idx + 2), self.get(idx + 1), self.get(idx)].into()
+    }
```
Doc comments for this function need to be updated.

Also, changing semantics in this way may lead to difficult-to-detect bugs downstream. At the very least, we should clearly document this as a breaking change in the changelog. But I also wonder if a different name should be used instead.

bobbinth commented on processor/src/stack/mod.rs:
```diff
@@ -161,13 +161,38 @@ impl Stack {
+    /// This method can access words that span into the overflow table.
+    /// Word 0 is defined by stack positions 0-3, word 1 by positions 4-7, etc.
+    /// The words are created in reverse order (top element at the last position).
+    pub fn get_word_with_overflow(&self, start_idx: usize) -> Word {
```
Similar comment as above - maybe there could be just a single function for getting words from the stack and it would handle both cases.

bobbinth commented on processor/src/host/default.rs:
```diff
@@ -154,11 +154,8 @@ where
-        event_id: u32,
-    ) -> Result<Vec<AdviceMutation>, EventError> {
+    fn on_event(&mut self, process: &ProcessState) -> Result<Vec<AdviceMutation>, EventError> {
+        let event_id = process.get_stack_item(0).as_int() as u32;
```
Similar to one of the other comments: I would rather use `try_into().unwrap()` than `as u32`.

bobbinth commented on processor/src/operations/sys_ops/sys_event_handlers.rs:
```diff
@@ -12,7 +12,8 @@ use miden_core::{
 
 /// The offset of the domain value on the stack in the `hdword_to_map_with_domain` system event.
-pub const HDWORD_TO_MAP_WITH_DOMAIN_DOMAIN_OFFSET: usize = 8;
+/// TODO: Mention we add 1 to account for event ID at the top
+pub const HDWORD_TO_MAP_WITH_DOMAIN_DOMAIN_OFFSET: usize = 9;
```
We should probably address this `TODO`.

bobbinth reviewed (commented):
Looks great! Thank you! I left some comments inline - but nothing major.

adr1anh commented on miden-vm/tests/integration/operations/decorators/mod.rs:
```diff
@@ -57,11 +57,8 @@ impl SyncHost for TestHost {
-        event_id: u32,
-    ) -> Result<Vec<AdviceMutation>, EventError> {
+    fn on_event(&mut self, process: &ProcessState) -> Result<Vec<AdviceMutation>, EventError> {
+        let event_id = process.get_stack_item(0).as_int() as u32;
```
Good catch, let me check the other places where we might be doing this conversion

adr1anh commented on docs/src/design/stack/system_ops.md:
```diff
@@ -11,15 +11,9 @@ s'_i - s_i = 0 \ \text{ for } i \in [0, 16) \text { | degree} = 1
->$$
-s'_i - s_i = 0 \ \text{ for } i \in [0, 16) \text { | degree} = 1
-$$
-
-Additionally, the prover puts `EMIT`'s immediate value in the first user op helper register non-deterministically. The [Op Group Table](../decoder/main.md#op-group-table) is responsible for ensuring that the prover sets the appropriate value.
+This operations therefore has the same semantics as the [`DROP`](./stack_ops.md#drop) operation for the stack.
```
I’m reworking all comments and will be pushing soon. But thanks for highlighting

bobbinth commented on docs/src/user_docs/assembly/events.md:
```diff
@@ -2,11 +2,13 @@
 Miden assembly supports the concept of events. Events are a simple data structure with a single `event_id` field.  When an event is emitted by a program, it is communicated to the host. Events can be emitted at specific points of program execution with the intent of triggering some action on the host. This is useful as the program has contextual information that would be challenging for the host  ...
 
-An event can be emitted via the `emit.<event_id>` assembly instruction where `<event_id>` can be any 32-bit value specified either directly or via a [named constant](./code_organization.md#constants). For example:
+An event can be emitted via the `emit.<event_id>` assembly instruction where `<event_id>` can be any 32-bit value specified either directly or via a [named constant](./code_organization.md#constants). Alternatively, the `emit` instruction can be used without an immediate value - in this case it will use the value from the top of the stack as the event ID. For example:
```
Ah! It was a typo on my end. I think I wanted to say is that we **should** allow full felt event IDs. Sorry!

adr1anh commented:
Apologies for the force push, due to the changes yesterday this was the easiest on my end. 

Thanks for the feedback @bobbinth , I've gone through all the documentation portion of the PR and addressed your comments.

The main breaking change was making `Stack::get` return elements from the overflow stack. This matches the behavior for the FastProcessor as well. Maybe @plafer has comments though.

bobbinth commented on core/src/operations/mod.rs:
```diff
@@ -169,7 +169,7 @@ pub enum Operation {
     /// Similar to Noop, this operation does not change the state of user stack. The immediate
     /// value affects the program MAST root computation.
-    Emit(u32) = OPCODE_EMIT,
+    Emit = OPCODE_EMIT,
```
I think lines 164 - 168 still need to be updated.

bobbinth commented on docs/src/user_docs/assembly/events.md:
```diff
@@ -1,13 +1,24 @@
+Events interrupt VM execution for one cycle and hand control to the host. The host can read VM state and modify the advice provider. From the VM's perspective, `emit` has identical semantics to `noop` - the operand stack and registers remain unchanged.
 
-An event can be emitted via the `emit.<event_id>` assembly instruction where `<event_id>` can be any 32-bit value specified either directly or via a [named constant](./code_organization.md#constants). For example:
+By convention, programs place a 32-bit `event_id` at the top of the stack before executing `emit`. The VM doesn't enforce this - the host reads and interprets the event ID.
```
I think we should remove the "32-bit" part from here (or replace it with "Felt").

bobbinth commented on docs/src/user_docs/assembly/instruction_reference.md:
```diff
@@ -307,10 +307,11 @@ High-level constructs for controlling the execution flow.
-| `trace.<trace_id>` | `[...]`     | `[...]`      | 0      | Emits a trace with the specified `trace_id` to the host. Does not change the state of the operand stack. The `trace_id` can be any 32-bit value specified either directly or via a [named constant](./code_organization.md#constants). Only active when programs are run with tracing flag (`-t` or `--trace`), otherwise ignored. Example: `trace ...
+| Instruction        | Stack Input       | Stack Output | Cycles | Notes                                                                                                                                                                                                                                                                                                                                        ...
+|--------------------|-------------------|--------------|--------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| `emit.<event_id>`  | `[...]`           | `[...]`      | 3      | Emits an event with the specified `event_id` to the host. The net effect on the operand stack is no change (internally expands to `push.<event_id> emit drop`). The `event_id` can be any 32-bit value specified either directly or via a [named constant](./code_organization.md#constants). Events allow programs to communicate contextua ...
```
Similar to the other comment - I'd remove the "32-bit" part (I don't think we are enforcing this, are we?).

bobbinth reviewed (approved):
Looks good! Thank you! I left a few small (mostly doc-related) comments inline.
